File: 백준/python/boj_2661.py
import sys

sys.stdin = open("input.txt", "r")

N = int(input())

isEnd = False

def isGood(num):
    for i in range(1, len(num)//2+1):
        if num[-i:] == num[-i*2: -i]:
            return False
    return True

def dfs(num):
    global isEnd
    if isEnd:
        return
    if len(num) == N:
        print(num)
        isEnd = True
    for i in range(1, 4):
        if isGood(num+str(i)):
            dfs(num+str(i))

dfs("")

File: 백준/js/boj_2661.js
const filePath = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(filePath).toString().trim();

const N = Number(input);

let isEnd = false;

function isGood(str) {
  let len = str.length;
  for (let i = 1; i <= len / 2; i++) {
    if (str.substr(len - i, i) === str.substr(len - i * 2, i)) {
      return false;
    }
  }
  return true;
}

function dfs(str) {
  if (isEnd) return;
  if (str.length == N) {
    console.log(str);
    isEnd = true;
  }
  for (let i = 1; i <= 3; i++) {
    let newStr = str + i.toString();
    if (isGood(newStr)) {
      dfs(newStr);
    }
  }
}

dfs("");
